### PR TITLE
Fix buffer overflow in Gorilla codec

### PR DESCRIPTION
**Pull request by evillique**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix crash due to buffer overflow while decompressing malformed data using `Gorilla` codec. This issue was found with [ClickHouse Bug Bounty Program](https://github.com/ClickHouse/ClickHouse/issues/38986) by https://twitter.com/malacupa

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit 533655a7ddfeb4409578e2406925a77620df726c with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57107/533655a7ddfeb4409578e2406925a77620df726c/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6959535165">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57107/533655a7ddfeb4409578e2406925a77620df726c/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Compression/CompressionCodecGorilla.cpp [hunk 1]
```diff
@@ -280,6 +281,9 @@ void decompressDataForType(const char * source, UInt32 source_size, char * dest)
     if (source + sizeof(T) > source_end || items_count < 1)
         return;
 
+    if (dest + items_count * sizeof(T) > dest_end)
```
Comment (alexey-milovidov) [inline src/Compression/CompressionCodecGorilla.cpp:284]: Is overflow possible here?

Comment (evillique) [reply src/Compression/CompressionCodecGorilla.cpp:284]: We can read any value into `items_count` from the data sent by the user, so the overflow seems to be easily possible.

Comment (alexey-milovidov) [reply src/Compression/CompressionCodecGorilla.cpp:284]: Ok, let's check for it with `arithmeticOverflow` or `__builtin_mul_overflow`

Comment (evillique) [reply src/Compression/CompressionCodecGorilla.cpp:284]: For example in the added test the `\xff\xff\xff\xff\` part is the value that would be read into `items_count`

Comment (evillique) [reply src/Compression/CompressionCodecGorilla.cpp:284]: Oh, the regular overflow

**Review by alexey-milovidov — APPROVED** [review]

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**